### PR TITLE
App shortcuts  & broadcasted events

### PR DESCRIPTION
**Pull request by jendakol**
This PR contains 2 features which solved or at least help with #149 :
1. Receive broadcasted events from other apps; can be easily used by Tasker, Automate, MacroDroid etc.
1. Handle app shortcuts - they can be called from automation apps too (not from all) but are also usable by the user directly, he can "lock" the shortcut to his "desktop" and use them like widgets

The original idea was to handle app shortcuts in a seamless way - just enabled/disable tracking, nothing more. Unfortunately, intents produced by shortcuts are limited to activity calling ones which obviously result in displaying the activity, no matter whether we want or don't want it. A solution may be to immediately call `finish()` and _hide_ the app again - however, this is not done quickly enough to be really smooth and it causes the app to "blink" on the foreground. Because of that, I've decided to add _synthetic_ delay before the app is hidden - so it shows up, the state is changed (sometimes it's changed even before it's shown up) and after 2 seconds, it's sent to the background.

What's to discuss is:
1. Name of the action
1. Type/data of the intent which is used in shortcuts (currently `RUN` with `Uri`). It's probably not a very interesting/important topic since this is only internal functionality - others should use the broadcasted event
1. Icons for shortcuts - see attachments
1. Delay before hiding the app while handling the shortcuts

Here comes examples of how the app shortcut looks like on Xiaomi A2 Lite (with a very colorful background) and on Samsung S10e (with a black background).  
I'm not sure if I'm happy with the appearance of Xiaomi, but in general, it depends on the background and also on specific phone manufacturers (it should be almost pure Android 9 in the case of Xiaomi A2 Lite, BTW).

IMG-20200415-WA0000
IMG-20200415-WA0001
Screenshot_20200415-230558_One UI Home
Screenshot_20200415-230609_One UI Home

### CODE REVIEW

File: app/src/main/kotlin/cz/covid19cz/erouska/receiver/ActionsReceiver.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    companion object {
+        private const val PREFIX = "EROUSKA"
+        const val ACTION_PAUSE = "${PREFIX}_PAUSE"
+        const val ACTION_RESUME = "${PREFIX}_RESUME"
+    }
+
+    override fun onReceive(context: Context, intent: Intent?) {
+        when (intent?.action) {
+            ACTION_PAUSE -> {
+                L.d("Received broadcasted intent to pause the scanning")
```
Comment (davidvavra) [inline app/src/main/kotlin/cz/covid19cz/erouska/receiver/ActionsReceiver.kt:19]: I would remove debug logs from the code

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 2]
```diff
@@ -1,36 +1,55 @@
...
 
+
 class MainActivity :
     BaseActivity<ActivityMainBinding, MainVM>(R.layout.activity_main, MainVM::class) {
 
+    companion object {
+        object ShortcutsActions {
+            private const val SCHEME = "erouska"
+
+            val URL_PAUSE: Uri = Uri.parse("${SCHEME}://${BuildConfig.FLAVOR}/pause")
```
Comment (davidvavra) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:44]: I would put flavor into scheme, better PROD will have only `erouska://` DEV will have `erouska-dev://`

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 3]
```diff
@@ -116,6 +144,31 @@ class MainActivity :
         customTabHelper.chromePackageName?.let {
             CustomTabsClient.bindCustomTabsService(this, it, customTabsConnection)
         }
+
```
Comment (davidvavra) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:147]: Why onStart and not OnCreate?

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 4]
```diff
@@ -85,7 +108,12 @@ class MainActivity :
             }
         })
 
-        viewModel.serviceRunning.value = CovidService.isRunning(this)
+        registerActionsReceiver()
```
Comment (davidvavra) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:111]: You also need to unregister receiver

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 5]
```diff
@@ -116,6 +144,31 @@ class MainActivity :
...
+            when (intent.data) {
+                ShortcutsActions.URL_PAUSE -> {
+                    L.d("Received action intent to pause the service")
+
+                    startService(CovidService.pause(this))
+
+                    Handler().postDelayed({
+                        L.d("Hiding the activity after handling of action intent")
+                        finishAfterTransition()
+                    }, 2000);
```
Comment (davidvavra) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:158]: I would put this magic number into a constant

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 6]
```diff
@@ -116,6 +144,31 @@ class MainActivity :
...
+                }
+                ShortcutsActions.URL_RESUME -> {
+                    L.d("Received action intent to resume the service")
+
+                    startService(CovidService.resume(this))
+
+                    Handler().postDelayed({
+                        L.d("Hiding the activity after handling of action intent")
+                        finishAfterTransition()
```
Comment (davidvavra) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:167]: Have you tested if you exit the activity by pressing back before the timeout ends?

File: app/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -33,6 +33,12 @@
                 <action android:name="android.intent.action.MAIN" />
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
+            <intent-filter>
```
Comment (davidvavra) [inline app/src/main/AndroidManifest.xml:36]: It should be documented how to use this "API" and app shortcuts. Maybe write a wiki page and link it from README.md?

File: app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -63,6 +69,7 @@
             </intent-filter>
         </receiver>
 
+        <receiver android:name=".receiver.ActionsReceiver"/>
```
Comment (simekadam) [inline app/src/main/AndroidManifest.xml:72]: The intent filter should be here in the manifest.  The way you did this the receiver can only be invoked when the MainActivity opens and in this state it's also effectively leaking the receiver.

File: app/src/main/kotlin/cz/covid19cz/erouska/receiver/ActionsReceiver.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import android.content.Context
+import android.content.Intent
+import cz.covid19cz.erouska.service.CovidService
+import cz.covid19cz.erouska.utils.L
+
+class ActionsReceiver : BroadcastReceiver() {
+    companion object {
+        private const val PREFIX = "EROUSKA"
+        const val ACTION_PAUSE = "${PREFIX}_PAUSE"
+        const val ACTION_RESUME = "${PREFIX}_RESUME"
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/receiver/ActionsReceiver.kt:13]: The actions should be prefixed with the package name and in lower case. It's not super important, but good practice when you expose it to others.

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 10]
```diff
@@ -116,6 +144,31 @@ class MainActivity :
...
         }
+
+        if (intent?.action == Intent.ACTION_RUN) {
+            when (intent.data) {
+                ShortcutsActions.URL_PAUSE -> {
+                    L.d("Received action intent to pause the service")
+
+                    startService(CovidService.pause(this))
+
+                    Handler().postDelayed({
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:155]: Not that stoked about this. I'd probably remove the autohide.
Have you tried to add new activity (that can be transparent? or in a form a of a dialog with simple UI) to handle the action?

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt (hunk 2 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:44]: Sorry, I forgot to mention this in "things to discuss" :-) I'll change the scheme and also the broadcasted intent.

File: app/src/main/AndroidManifest.xml (hunk 7 above)
Comment (jendakol) [reply app/src/main/AndroidManifest.xml:36]: Sure, just didn't want to go too much ahead :-D

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 11]
```diff
@@ -162,4 +215,48 @@ class MainActivity :
         return bluetoothRepository.isBtEnabled() && isLocationEnabled() && hasLocationPermission()
     }
 
+    private fun registerActionsReceiver() {
+        val actionsFilter = IntentFilter(ActionsReceiver.ACTION_PAUSE)
+        actionsFilter.addAction(ActionsReceiver.ACTION_RESUME)
+        registerReceiver(actionsReceiver, actionsFilter)
+    }
```
[the remark this answers is not in the document]
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:222]: I don't understand. This has to be here...

File: app/src/main/kotlin/cz/covid19cz/erouska/receiver/ActionsReceiver.kt (hunk 9 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/receiver/ActionsReceiver.kt:13]: Is it? I didn't want to make it too long and hard to write. With the prefix, it seemed still unique enough to me :thinking:

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt (hunk 10 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:155]: I've tried adding new activity. The problem with blinking was the same. But having a transparent activity - nope, I didn't try to do that :smile: I'll give it a try.

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 12]
```diff
@@ -162,4 +215,48 @@ class MainActivity :
...
+
+    private fun manageShortcuts(isRunning: Boolean) {
```
[the remark this answers is not in the document]
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:224]: Well, OK, I'll try.

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (simekadam) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:222]: This way the receiver only works when the app is running.
You should define the intent filter in manifest inside of the BR declaration and then Android will be able to wake up the app even when it's not running.

[the remark this answers is not in the document]
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:222]: You mean the filtering? I got inspired by other receivers :-P

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt (hunk 5 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:158]: Removed

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt (hunk 6 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:167]: Timeout removed

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/ShortcutActivity.kt [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+class ShortcutActivity : AppCompatActivity() {
+
+    private val shortcutsManager = ShortcutsManager(this)
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        intent.let {
+            shortcutsManager.handlePossibleShortcut(it)
+        }
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/ShortcutActivity.kt:16]: shortcutsManager.handlePossibleShortcut(intent) nebo intent?.let{ … } ?

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 14]
```diff
@@ -25,13 +25,17 @@ import cz.covid19cz.erouska.utils.CustomTabHelper
...
 
+
 class MainActivity :
     BaseActivity<ActivityMainBinding, MainVM>(R.layout.activity_main, MainVM::class) {
 
     private val localBroadcastManager by inject<LocalBroadcastManager>()
     private val bluetoothRepository by inject<BluetoothRepository>()
     private val customTabHelper by inject<CustomTabHelper>()
 
+    //    private val actionsReceiver by inject<ActionsReceiver>()
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:36]: 🗑️

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/ShortcutsManager.kt [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    companion object {
+        object ShortcutsActions {
+            private val SCHEME = "erouska" + if (BuildConfig.FLAVOR == "dev") "-dev" else ""
+
+            val URL_PAUSE: Uri = Uri.parse("${SCHEME}://service/pause")
+            val URL_RESUME: Uri = Uri.parse("${SCHEME}://service/resume")
+        }
+    }
+
+    fun handlePossibleShortcut(intent: Intent) {
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/ShortcutsManager.kt:24]: NIT rename to just handleShortcut? Or handleShortcutIntent

File: app/src/main/kotlin/cz/covid19cz/erouska/DI.kt [hunk 16]
```diff
@@ -84,6 +85,7 @@ val repositoryModule = module {
 val appModule = module {
     single { LocationStateReceiver() }
     single { BluetoothStateReceiver() }
+    single { ActionsReceiver() }
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/DI.kt:88]: 🗑️ ?

File: app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt [hunk 17]
```diff
@@ -160,11 +163,14 @@ class CovidService : Service() {
         prefs.setAppPaused(false)
         createNotification()
         turnMaskOn()
+        shortcutsManager.updateShortcuts(!servicePaused)
     }
 
     private fun stop(intent: Intent) {
         wakeLockManager.release()
         servicePaused = true
+        shortcutsManager.updateShortcuts(!servicePaused)
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt:172]: NIT extract to method?

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt [hunk 18]
```diff
@@ -85,7 +92,10 @@ class MainActivity :
             }
         })
 
-        viewModel.serviceRunning.value = CovidService.isRunning(this)
+        val isRunning = CovidService.isRunning(this)
+
+        viewModel.serviceRunning.value = isRunning
+        shortcutsManager.updateShortcuts(isRunning)
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:98]: Do you also need to update shortcuts from here when the service does it?

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/ShortcutsManager.kt [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+                .setIcon(IconCompat.createWithResource(context, R.drawable.ic_shortcut_pause))
+                .setIntent(
+                    Intent(
+                        Intent.ACTION_RUN,
+                        ShortcutsActions.URL_PAUSE
+                    ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
+                )
+                .build()
+
+            listOf(shortcut)
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/ui/main/ShortcutsManager.kt:53]: NIT, you can return just the one shortcut from here and wrap it in that list below

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt (hunk 18 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:98]: I do. I noticed that when the scanning is paused and the app starts, the service is not started therefore `updateShortcuts` is not called.

Comment (simekadam) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:97]: But the shortcut should be in the correct state already, shouldn't it?

File: app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt [hunk 20]
```diff
@@ -120,6 +121,7 @@ class CovidService : Service() {
     override fun onCreate() {
         super.onCreate()
         subscribeToReceivers()
+        updateAppShortcuts()
```
Comment (simekadam) [inline app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt:124]: One thing I just realized is that you should update the shortcut from onDestroy. 
I'm wondering what;s gonna happen when the app is force closed. We can also tweak this later

File: app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt (hunk 18 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/ui/main/MainActivity.kt:97]: Well, in theory...

File: app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt (hunk 20 above)
Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt:124]: I can try, but deadline from @davidvavra has just been crossed :smile:

Comment (simekadam) [reply app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt:124]: Lol, :shipit:

Comment (jendakol) [reply app/src/main/kotlin/cz/covid19cz/erouska/service/CovidService.kt:124]: Unfortunately, the shortcut is not refreshed even when I try to do it in `onDestroy`. However, this is so big corner-case, I would just ignore it. One must count with _werid behaviors when killing some app, right?

File: README.md [hunk 21]
```diff
@@ -89,4 +89,17 @@ eRouška uses GitHub Actions. A push to master branch triggers an App build. The
...
+As was proposed in [#149](https://github.com/covid19cz/erouska-android/issues/149), the app now supports
+controlling from other apps by sending broadcasted intents.  
+In your favorite automation app (Automate, Tasker, MacroDroid, ...) it should be enough to use an intent
+broadcast with these params:
+
+|Name|Value|
+|----|-----|
+|Package|`cz.covid19cz.erouska`|
+|Receiver/class|`cz.covid19cz.erouska.receiver.ActionsReceiver`|
+|Action|`cz.covid19cz.erouska.EROUSKA_PAUSE`/`cz.covid19cz.erouska.EROUSKA_PAUSE`|
```
Comment (davidvavra) [inline README.md:105]: Tady je dvakrát pause

**Pull request merged**